cluster_experiments/analysis_plan.py

```diff
@@ -0,0 +1,264 @@
+        ValueError
+            If tests or variants are empty lists.
+        """
+        if not isinstance(tests, list) or not all(
+            isinstance(test, HypothesisTest) for test in tests
+        ):
+            raise TypeError("Tests must be a list of HypothesisTest instances")
+        if not isinstance(variants, list) or not all(
+            isinstance(variant, Variant) for variant in variants
+        ):
+            raise TypeError("Variants must be a list of Variant instances")
+        if not isinstance(variant_col, str):
+            raise TypeError("Variant_col must be a string")
+        if not tests:
+            raise ValueError("Tests list cannot be empty")
+        if not variants:
+            raise ValueError("Variants list cannot be empty")
+
+    def analyze(
+        self,
+        exp_data: pd.DataFrame,  # , pre_exp_data: Optional[pd.DataFrame], alpha=0.05
+    ) -> DataFrame:
+        ...
+        # add methods to prepare the filtered dataset based on variants and slicers
+        # add methods to run the analysis for each of the hypothesis tests, given a filtered dataset
+        # store each row as a HypothesisTestResults object
+        # wrap all results in an AnalysisPlanResults object
+
+        # -----
+
+        # add all kind of checks on the inputs at the beginning using the data structures
+        # todo: ...
+        # do it before running the computations below
+
+        test_results = []
+        treatment_variants: List[Variant] = self.get_treatment_variants()
+        control_variant: Variant = self.get_control_variant()
+
+        for test in self.tests:
+            # todo: add cupac handler here
+            analysis_class = analysis_mapping[test.analysis_type]
+            target_col = test.metric.get_target_column_from_metric()
+
+            for treatment_variant in treatment_variants:
+
+                analysis_config_final = self.prepare_analysis_config(
+                    initial_analysis_config=test.analysis_config,
+                    target_col=target_col,
+                    treatment_col=self.variant_col,
+                    treatment=treatment_variant.name,
+                )
+
+                experiment_analysis = analysis_class(**analysis_config_final)
+
+                for dimension in test.dimensions:
+                    for dimension_value in list(set(dimension.values)):
+                        prepared_df = self.prepare_data(
+                            data=exp_data,
+                            variant_col=self.variant_col,
+                            treatment_variant=treatment_variant,
+                            control_variant=control_variant,
+                            dimension_name=dimension.name,
+                            dimension_value=dimension_value,
+                        )
+
+                        ate = experiment_analysis.get_point_estimate(df=prepared_df)
```

Comment:
I think I'd send the responisibility of preparing ate, pvalue, std error and confidence interval into a single object to experimentanalysis

Reply:
makes sense! will do ;) I already had this in mind to not repeat the fit of the models multiple times